k8s/cmd/commands/sidecar/inject.go

```diff
@@ -0,0 +1,313 @@
+
+	sidecar, err := buildSidecar(namespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build sidecar: %w", err)
+	}
+
+	manifests := []string{}
+	yamls := []string{}
+	t := templates.NewEmbedTemplate()
+	etcdManifests, err := t.GetFilesRecursive("etcd")
+	if err != nil {
+		return manifests, fmt.Errorf("failed to get etcd manifests: %v", err)
+	}
+	sidecarManifests, err := t.GetFilesRecursive("sidecar")
+	if err != nil {
+		return manifests, fmt.Errorf("failed to get sidecar manifests: %v", err)
+	}
+
+	vineyardManifests, err := t.GetFilesRecursive("vineyardd")
+	if err != nil {
+		return manifests, fmt.Errorf("failed to get vineyardd manifests: %v", err)
+	}
+
+	tmplFunc := map[string]interface{}{
+		"getEtcdConfig":           getEtcdConfig,
+		"getServiceLabelSelector": getSidecarLabelSelector,
+	}
+
+	// set up the etcd
+	EtcdConfig.Namespace = namespace
+	etcdEndpoints := make([]string, 0, sidecar.Spec.Replicas)
+	replicas := sidecar.Spec.Replicas
+	for i := 0; i < replicas; i++ {
+		etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("etcd%v=http://etcd%v:2380", strconv.Itoa(i), strconv.Itoa(i)))
+	}
+	EtcdConfig.Endpoints = strings.Join(etcdEndpoints, ",")
+	// the etcd is built in the vineyardd image
+	EtcdConfig.Image = sidecar.Spec.VineyardConfig.Image
+	for i := 0; i < replicas; i++ {
+		EtcdConfig.Rank = i
+		for _, ef := range etcdManifests {
+			manifest, err := t.ReadFile(ef)
+			if err != nil {
+				return manifests, fmt.Errorf("failed to read manifest %s: %v", ef, err)
+			}
+			manifests = append(manifests, string(manifest))
+		}
+	}
+
+	// set up the service for etcd
+	for _, vf := range vineyardManifests {
+		if vf == "vineyardd/etcd-service.yaml" || vf == "vineyardd/service.yaml" {
+			manifest, err := t.ReadFile(vf)
+			if err != nil {
+				return manifests, fmt.Errorf("failed to read manifest %s: %v", vf, err)
+			}
+			manifests = append(manifests, string(manifest))
+		}
+	}
+
+	// set up the vineyard sidecar
+	for _, sf := range sidecarManifests {
+		manifest, err := t.ReadFile(sf)
+		if err != nil {
+			return manifests, fmt.Errorf("failed to read manifest %s: %v", sf, err)
+		}
+		manifests = append(manifests, string(manifest))
+	}
+
+	for _, m := range manifests {
+		obj := &unstructured.Unstructured{}
+		_, _ = swckkube.LoadTemplate(m, sidecar, tmplFunc, obj)
+		if obj.GetKind() != "" {
+			if obj.GetName() == "injection-containers" {
+				if err := InjectSidecarConfig(workloadObj, obj); err != nil {
+					return manifests, fmt.Errorf("failed to inject the sidecar config: %w", err)
+				}
+				obj = workloadObj
+			}
+
+			ss, err := obj.MarshalJSON()
+			if err != nil {
+				return manifests, fmt.Errorf("failed to marshal the unstructuredObj: %w", err)
+			}
+
+			yaml, err := util.ConvertToYaml(string(ss))
+			if err != nil {
+				return manifests, fmt.Errorf("failed to convert the unstructuredObj to yaml: %w", err)
+			}
+			yamls = append(yamls, yaml)
+		}
+	}
+
+	return yamls, nil
```

Comment:
I cannot remember in which file I have meet this block, but it looks quite familiar. Maybe you could take a double-check to find if it can reuse some existing functions?